### PR TITLE
Implement the JVM chaos controller

### PR DESCRIPTION
**Pull request by Gallardot**
### What problem does this PR solve?
#650 

### What is changed and how does it work?
We use [chaosblade-exec-jvm](https://github.com/chaosblade-io/chaosblade-exec-jvm) to implement JVM chaos. We use  admission-webhook to inject the required files and associated configurations. We use [JAVA_TOOL_OPTIONS](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/envvars002.html) to start a java agent with application.

ps: It is currently only a technical preview, and we need to do more for the documentation and E2E test.  😂

Here are the steps to perform JVM chaos. Have fun.🚀

```
# create app namespace
kubectl create namespace app

// enable chaos mesh admission-webhook
kubectl label namespace app admission-webhook=enabled

// add the required sidecar profile
kubectl apply -f  examples/jvm/sidecar.yaml  
kubectl apply -f  examples/jvm/sidecar-template.yaml 
```

Preparation done
At this point, we need to prepare a Java application pod, adding the following information to the annotations of pod
```
admission-webhook.chaos-mesh.org/request: jvmchaos-sidecar
```

As an example, you can use our [demo program](https://github.com/chaos-mesh/jvm-chaos-demo). 
```
// deploy a demo app
kubectl apply -f examples/jvm/app.yaml

// wait until the pod is running
kubectl get pods -n app


```

JVM Agent injection validation
```
// Open two new terminals to perform the local service
kubectl port-forward -n app deployment/springboot-jvmchaos-demo 12345:8080
kubectl port-forward -n app deployment/springboot-jvmchaos-demo 10086:10086

// Check that the JVM Agent has been injected
curl -N -s http://127.0.0.1:10086/sandbox/default/module/http/sandbox-module-mgr/list

// response information is as follows
sandbox-info        	ACTIVE  	LOADED  	0    	0    	0.0.4          	luanjia@taobao.com
sandbox-module-mgr  	ACTIVE  	LOADED  	0    	0    	0.0.2          	luanjia@taobao.com
chaosblade          	FROZEN  	LOADED  	0    	0    	0.9.0          	Changjun Xiao
sandbox-control     	ACTIVE  	LOADED  	0    	0    	0.0.3          	luanjia@taobao.com
total=4
```

make a delayed fault injection test
```
// Very fast response
curl -w "time_total:%{time_total}s" http://127.0.0.1:12345/hello

// Starts a delayed fault injection for all path in the servlet
kubectl apply -f examples/jvm/jvm-delay4servlet-example.yaml 

// about 10s delay
curl -w "time_total:%{time_total}s" http://127.0.0.1:12345/hello

// clean 
 kubectl delete -f examples/jvm/jvm-delay4servlet-example.yaml 

// Very fast response
curl -w "time_total:%{time_total}s" http://127.0.0.1:12345/hello
```

Do a test to modify the method's return value
```
// return `hello world!`
curl http://127.0.0.1:12345/hello

// change the return value to 'hello chaos mesh!'
kubectl apply -f examples/jvm/jvm-return-example.yaml 

// return `hello chaos mesh!`
curl http://127.0.0.1:12345/hello

// clean 
kubectl delete -f examples/jvm/jvm-return-example.yaml 

// return `hello world!`
curl http://127.0.0.1:12345/hello
``` 

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Breaking backward compatibility

Related changes

- [ ] Need to update the documentation

### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by STRRL** [comment]
Hi @Gallardot, is there any documents about the web API `http://%s:%d/sandbox/default/module/http/chaosblade/create` and docker image `gallardot/chaos-jvm`?😁

**Comment by Gallardot** [comment]
> Hi @Gallardot, is there any documents about the web API `http://%s:%d/sandbox/default/module/http/chaosblade/create` and docker image `gallardot/chaos-jvm`?😁

Sorry, the web API has no documentation. But we can find  [the implementation code](https://github.com/chaosblade-io/chaosblade-exec-jvm/tree/master/chaosblade-exec-service/src/main/java/com/alibaba/chaosblade/exec/service/handler).

The docker image `gallardot/chaos-jvm` is our [demo program](https://github.com/chaos-mesh/jvm-chaos-demo).

**Comment by STRRL** [comment]
Hi @Gallardot , sorry for my late reply, as you mentioned before, copy "jvm-sandbox" then attach JVM in the container works well. 

I think injecting chaos at runtime is better than manually changing workload spec, although the first one is more complex.

How about making `JAVA_HOME` and `USER` (and other necessary things) as fields in the JVMChaosSpec? As we have "chaos-daemon", we could implement "copy file" and "change user" without kubernetes API.

Multi-container and more than one java process should also be considered, we should enrich out chaos spec.

**Comment by Gallardot** [comment]
As we discussed, executing JVM Chaos by Chaos-daemon is desirable, but difficult to handle all cases. With Sidecar we can handle most cases.
Therefore, we plan to implement both ways, and the user can choose one or the other depending on the actual situation. @STRRL  is going to help us implement the Chaos-daemon way. Thanks to @STRRL .❤️

We want to merge the current PR first, and then we create other PR to implement the Chaos-daemon way and supplement the E2E use cases and documentation.

ps: I will fix the conflicting files later.

@cwen0 @YangKeao   PTAL

File: pkg/jvm/action.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	SUID   string = "suid"
+	ACTION string = "action"
+	TARGET string = "target"
+)
+
+// ToSandboxAction convertes chaos to sandbox action
+func ToSandboxAction(suid string, chaos *v1alpha1.JVMChaos) ([]byte, error) {
+	kv := make(map[string]string, 0)
+
+	flags := v1alpha1.JvmSpec[chaos.Spec.Target][chaos.Spec.Action].Flags
```
Comment (STRRL) [inline pkg/jvm/action.go:34]: `JvmSpec` is two-level map, we need to avoid nil for each layer.

File: pkg/jvm/action.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+					f, _ := strconv.ParseBool(v)
+					if f {
+						kv[k] = v
+					}
+				}
+			}
+		}
+	}
+
+	matchers := v1alpha1.JvmSpec[chaos.Spec.Target][chaos.Spec.Action].Matchers
```
Comment (STRRL) [inline pkg/jvm/action.go:54]: Same issue with two-level map.

File: api/v1alpha1/jvmchaos_webhook.go [hunk 3]
```diff
@@ -76,3 +85,557 @@ func (in *JVMChaos) ValidateScheduler(spec *field.Path) field.ErrorList {
...
+	}
+	return allErrs
+}
+
+// +kubebuilder:object:generate=false
+
+// JvmSpec from chaosblade-jvm-spec.yaml
+// chaosblade-jvm-spec.yaml file generated by
+// https://github.com/chaosblade-io/chaosblade-exec-jvm/blob/master/chaosblade-exec-service/src/main/java/com/alibaba/chaosblade/exec/service/build/SpecMain.java
+var JvmSpec = map[JVMChaosTarget]map[JVMChaosAction]ActionParameterRules{
```
Comment (STRRL) [inline api/v1alpha1/jvmchaos_webhook.go:178]: Does `JvmSpec` need manual maintenance? Any idea to automate it?

**Review by STRRL — CHANGES_REQUESTED** [review]
Thanks for your contribution! 
The API is much clear than the last verison.😁 

PTAL @Gallardot

File: pkg/jvm/action.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+					continue
+				}
+
+				if rule.ParameterType != v1alpha1.BoolType {
+					kv[k] = v
+				} else {
+					f, _ := strconv.ParseBool(v)
+					if f {
+						kv[k] = v
+					}
```
Comment (STRRL) [inline pkg/jvm/action.go:48]: This error is omitted. I think it's better to return this error if exists, then setup `kv[k] = v` explicitly.

Same issue at lines 65 - 69.

File: api/v1alpha1/jvmchaos_webhook.go (hunk 3 above)
Comment (Gallardot) [reply api/v1alpha1/jvmchaos_webhook.go:178]: @STRRL 

Yes,if there is an update or version upgrade to the chaosBlade -jvm-spec.yaml file, we may need to update the JVMSpec. There are partial differences between the definition of JVMSpec and that of chaosBlade -jvm-spec.yaml. The main difference is in the type definition.  A clearer type definition would help us do some data validation ahead of time. The difference is artificial, and there may be some specific criteria. ChaosBlade -exec-JVM does not have frequent version updates, so I find it acceptable to maintain it manually for now.

File: pkg/jvm/action.go (hunk 2 above)
Comment (Gallardot) [reply pkg/jvm/action.go:66]: All done. PTAL

File: api/v1alpha1/jvmchaos_webhook.go (hunk 3 above)
Comment (STRRL) [reply api/v1alpha1/jvmchaos_webhook.go:178]: Oh, I got that.

**Review by STRRL — COMMENTED** [review]
rest LGTM, PTAL @Gallardot

File: pkg/jvm/action.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+				if rule.ParameterType != v1alpha1.BoolType {
+					kv[k] = v
+				} else {
+					f, err := strconv.ParseBool(v)
+					if err != nil {
+						return nil, fmt.Errorf("can not parse Spec.Flags.%s's value:%s as boolean", k, v)
+					}
+					if f {
+						kv[k] = v
+					}
```
[the remark this answers is not in the document]
Comment (Gallardot) [reply pkg/jvm/action.go:60]: chaos-blade  just determines whether the key exists, not the value. If this `f` is false, we should not send the `key` to chaos-blade

File: pkg/jvm/sandbox.go [hunk 6]
```diff
@@ -39,26 +39,19 @@ func ActiveSandbox(host string, port int) error {
...
-	client := &http.Client{}
-	reqBody := bytes.NewBuffer([]byte(body))
 	url := fmt.Sprintf(RecoverURL, host, port)
+	return httpPost(url, body)
+}
 
+func httpPost(url string, body []byte) error {
+	client := &http.Client{}
+	reqBody := bytes.NewBuffer([]byte(body))
 	request, _ := http.NewRequest("POST", url, reqBody)
```
[the remark this answers is not in the document]
Comment (Gallardot) [reply pkg/jvm/sandbox.go:56]: This is just building the Request object, not sending an HTTP Request. 

```
response, _ := client.Do(request)
```
Maybe this part you're talking about. I have fix it. PTAL, Thanks!

[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/jvm/sandbox.go:56]: Oh, I think we should accept 2xx not only 200. :P @Gallardot

File: pkg/jvm/action.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/jvm/action.go:60]: Oh, I got that. 🤪 
How about appending comments about these "flag-style" configurations for `boolean`.

File: pkg/jvm/sandbox.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Gallardot) [reply pkg/jvm/sandbox.go:56]: I think chaosblade-exec-jvm is a simple http server. Just a few simple HTTP response codes. I think 200 and non-200  is enough. @STRRL

[the remark this answers is not in the document]
Comment (STRRL) [reply pkg/jvm/sandbox.go:56]: OK!

**Review by STRRL — APPROVED** [review]
LGTM

**Review by cwen0 — COMMENTED** [review]
Rest LGTM

File: controllers/jvmchaos/types.go [hunk 7]
```diff
@@ -31,13 +46,171 @@ type endpoint struct {
...
+		return err
+	}
+	return nil
+}
+
+func (r *endpoint) applyPod(ctx context.Context, pod *v1.Pod, chaos *v1alpha1.JVMChaos) error {
+	r.Log.Info("Try to apply jvm chaos", "namespace",
+		pod.Namespace, "name", pod.Name)
+
+	err := jvm.ActiveSandbox(pod.Status.PodIP, 10086)
```
[the remark this answers is not in the document]
Comment (Gallardot) [reply controllers/jvmchaos/types.go]: Done!  PTAL!

**Review by STRRL — DISMISSED** [review]
LGTM

**Review by cwen0 — DISMISSED** [review]
LGTM

**Review by cwen0 — APPROVED** [review]

**Review by STRRL — APPROVED** [review]

**Pull request merged**